cupyx/scipy/sparse/_index.py

```diff
@@ -0,0 +1,388 @@
+        elif isinstance(row, slice):
+            if isinstance(col, _int_scalar_types):
+                return self._get_sliceXint(row, col)
+            elif isinstance(col, slice):
+                if row == slice(None) and row == col:
+                    return self.copy()
+                return self._get_sliceXslice(row, col)
+            elif col.ndim == 1:
+                return self._get_sliceXarray(row, col)
+            raise IndexError('index results in >2 dimensions')
+        elif row.ndim == 1:
+            if isinstance(col, _int_scalar_types):
+                return self._get_arrayXint(row, col)
+            elif isinstance(col, slice):
+                return self._get_arrayXslice(row, col)
+        else:  # row.ndim == 2
+            if isinstance(col, _int_scalar_types):
+                return self._get_arrayXint(row, col)
+            elif isinstance(col, slice):
+                raise IndexError('index results in >2 dimensions')
+            elif row.shape[1] == 1 and (col.ndim == 1 or col.shape[0] == 1):
+                # special case for outer indexing
+                return self._get_columnXarray(row[:, 0], col.ravel())
+
+        # The only remaining case is inner (fancy) indexing
+        row, col = cupy.broadcast_arrays(row, col)
+        if row.shape != col.shape:
+            raise IndexError('number of row and column indices differ')
+        if row.size == 0:
+            return self.__class__(cupy.atleast_2d(row).shape, dtype=self.dtype)
+        return self._get_arrayXarray(row, col)
+
+    def _parse_indices(self, key):
+        M, N = self.shape
+        row, col = _unpack_index(key)
+
+        # Scipy calls sputils.isintlike() rather than
+        # isinstance(x, _int_scalar_types). Comparing directly to int
+        # here to minimize the impact of nested exception catching
+
+        if isinstance(row, _int_scalar_types):
+            row = _normalize_index(row, M, 'row')
+        elif not isinstance(row, slice):
+            row = self._asindices(row, M)
+
+        if isinstance(col, _int_scalar_types):
+            col = _normalize_index(col, N, 'column')
+        elif not isinstance(col, slice):
+            col = self._asindices(col, N)
+
+        return row, col
+
+    def _asindices(self, idx, length):
+        """Convert `idx` to a valid index for an axis with a given length.
+        Subclasses that need special validation can override this method.
+        """
+        try:
+            x = cupy.asarray(idx, dtype=self.indices.dtype)
+        except (ValueError, TypeError, MemoryError):
+            raise IndexError('invalid index')
+
+        if x.ndim not in (1, 2):
+            raise IndexError('Index dimension must be <= 2')
+
+        if x.size == 0:
+            return x
+
+        if x[x < 0].size > 0:
+            if x is idx or not x.flags.owndata:
+                x = x.copy()
+            x[x < 0] += length
+        return x
```

Comment:
How about the following case for the simplification?
```suggestion
        return x % length
```

Reply:
This one's actually a little hard to test right now because this is scaffolding for fancy/advanced indexing which isn't supported yet in this PR (and thus, untested in this PR.) This code is taken almost directly from Scipy, though, and it there are tests in the subsequent PRs which test it. 

Since `_asindices` is not yet called in this PR, what do you think about leaving this as-is just for this PR and I can simplify the function in the follow-on PR that includes advanced indexing so that it can be tested? It would just be unfortunate to make a breaking change like this and find out in the follow-on PRs that something broke.

Reply:
OK. I'll re-request it in the follow-on PR.

Reply:
I just made this change in https://github.com/cupy/cupy/pull/3658 and verified it passes all the tests.